### ISSUE TITLE
Dirty Hive support

### ISSUE THREAD

**Issue by muteb** [janstarke/nt-hive2#1]
Hi,
Thank you very much for the awesome lib.. 
From a forensics prospective, any plan to support rebuilding dirty hives by merging LOG1 or LOG2? If not, I might have something already done on RUST with the help of "[forensicmatt/r-winreg](https://github.com/forensicmatt/r-winreg)" earlier when I started learning rust and I will do my best to push it soon.
BTW, my code is not as good as yours as but I will try to be as consistent as I can with your code flow .

**Comment by janstarke** [comment]
That would be a very helpful feature. I'm looking forward to your contribution

**Comment by muteb** [comment]
Hi Jan,
I'm working on it and hopefully the next few coming days will be pushed.I'm struggling a bit with binread but catching up. One thing that I spent so much time is how would you do this function with the help of binread?

fn compute_checksum(buffer: &[u8;4096])->u32 {
        let mut chk: u32 = 0;
        // compute over first 508 bytes of block
        for i in 0..127 {
            let pos = i * 4;
            let chunk = LittleEndian::read_u32(&buffer[pos..pos + 4]);
            chk ^= chunk;
        }
        chk
    }
    
    1. how can you specify the size [u8;4096] with the help of binread?
    2. what would be the possible way to convert this "LittleEndian::read_u32(&buffer[pos..pos + 4]);" to binread

**Comment by muteb** [comment]
Hi Janstarke,
I'm %99 almost done but I need to verify the above question as well as once the recovery is done, should I extract the new clean file and save it to a directory or return it as an address and keep it in the memory?

**Comment by janstarke** [comment]
I added a sample checksum function for the hive base block in hive.rs , as I could imagine how a checksum could be calculated using binread:

```rust
    fn validate_checksum(data: &mut B) -> BinResult<()> {
        let current_position = data.stream_position()?;
        let  _: HiveBaseBlockRaw = match data.read_le() {
            Ok(r) => r,
            Err(why) => {
                log::error!("invalid checksum detected");
                return Err(why);
            }
        };
        data.seek(SeekFrom::Start(current_position))?;
        Ok(())
    }
```
```rust
#[allow(dead_code)]
#[derive(BinRead)]
struct HiveBaseBlockRaw {
    #[br(count = 127)]
    raw_data: Vec<u32>,

    #[br(assert(checksum_of(&raw_data[..]) == checksum))]
    checksum: u32
}

/// <https://systemroot.gitee.io/pages/apiexplorer/d9/d1/hivesum_8c.html#a0>
fn checksum_of(bytes: &[u32]) -> u32 {
    match bytes.iter().fold(0, |acc, x| acc ^ x) {
        0xffff_ffff => 0xffff_fffe,
        0           => 1,
        sum    => sum
    }
}
```

This might not be the best approach, but one how I would implement this

**Comment by muteb** [comment]
Noted. Thanks alot

**Comment by janstarke** [comment]
Do you have any good sources of information where I can learn something about dirty hive files?

**Comment by muteb** [comment]
The main resource I relay on is "https://github.com/msuhanov/regf/blob/master/Windows%20registry%20file%20format%20specification.md#dirty-state-of-a-hive" and then I stuck a bit with which log to be applied first and I found this code "https://github.dev/EricZimmerman/Registry/blob/fc8a166c812144e97f197670b7a9d35f3d276f22/Registry/Other/RegistryHeader.cs#L160" very helpful.

**Comment by janstarke** [comment]
I started a first additional binary `cleanhive` which evolved from your sample. I still have it in the feature branch as development state only, until we've solved the license problem.

**Comment by muteb** [comment]
Oh cool. I just checked it out and thank you very much

**Comment by muteb** [comment]
Hi jan,
I have rewritten the Marvin part a fixed some issues and I will push it soon but this might be a noob question as I never dealt with branches. my question is how can I submit the changes to the branch?

Thanks,

**Comment by muteb** [comment]
Never mind, I pushed it :).

**Comment by janstarke** [comment]
There are some merge conflicts, which I cannot resolve because I don't have write access to muteb:master. Could you please resolve them?

**Comment by muteb** [comment]
Hi Jan,
I might have accidentally merged  my master to your repo's branch and didn't wait for your approval. Please have a look at it and I'm sorry for the mistake. Every things works as expected with this version and solved multiple things mentioned on the comments earlier. I will have another fix by the end of this week trying to avoid the unwrap and make shorter functions if possible.
